tap_autopilot/__init__.py

```diff
@@ -195,19 +187,14 @@ def gen_request(STATE, endpoint, params=None):
     with metrics.record_counter(source) as counter:
         while True:
             data = request(endpoint, params).json()
-            if 'contact' in source:
-                if "bookmark" in data:
-                    params["bookmark"] = data["bookmark"]
-
-                else:
-                    params = {}
 
             for row in data[source_key]:
                 counter.increment()
                 yield row
 
-            if len(data[source_key]) < PER_PAGE:
-                params = {}
+            if "bookmark" in data:
+                params["bookmark"] = data["bookmark"]
+            else:
                 break
```

Comment:
It's not a blocker for the PR, but it looks like if there's only one condition in this while loop can it be refactored so that instead of
```
while True:
    if condition:
       do something
    else:
       break
```
It somehow is written as:
```
while condition:
    do something
```
?

Reply:
The `if/else` condition in the `while` loop checks state that isn't available until after the first statement in the loop is executed. Since Python doesn't provide `do/while`, seems like the only alternative to `while True` would be to initialize something before the start of the loop, like `paginate = True` or `bookmark = ""`. To me this seems like just introducing more complexity, but I'm happy to make a change in that direction if it's more inline with internal style guidelines! :)

Reply:
I mean, you could, as you suggest, initialize `bookmark = "start_loop"` or any string that is not blank (since blank evaluates to false) always set bookmark to equal `data["bookmark"]` inside the loop and then it should work. You don't need to conditionally modify the variable. Then set params["bookmark"] to equal the bookmark variable at the end of the loop, every time You don't need an if condition because params is not used after the while loop exits (which if I'm reading correctly is the last line of that function).

What do you think? Personally I think it is cleaner with less conditions, which means less branching, less to keep in my head as I'm reading the code.

Reply:
Is this what you have in mind?
```python
        bookmark = ""
        while bookmark is not None:
            data = request(endpoint, params).json()

            for row in data[source_key]:
                counter.increment()
                yield row

            bookmark = data.get("bookmark")
            params["bookmark"] = bookmark
```

Reply:
Oh, also, you can do `bookmark = data.get("bookmark", None)` to avoid any failure in the retrieval if the value is missing from the json.

Reply:
The dictionary `get` method returns `None` by default when the key is missing, so `data.get("bookmark")` is equivalent to `data.get("bookmark", None)`, but I will add the explicit default to avoid any confusion
```
In [1]: data = {}

In [2]: data.get("bookmark") == data.get("bookmark", None) == None
Out[2]: True
```

Reply:
@asaf-erlich changes pushed -- let me know if there's anything else you need to approve the PR

Reply:
Interesting, I didn't know .get returned None, I thought it failed if it was missing in the same way ['bookmark'] would, but that works. Thanks for making the changes.